sql/core/src/test/scala/org/apache/spark/sql/jdbc/DockerHacks.scala

```diff
@@ -0,0 +1,166 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.jdbc
+
+import java.sql.Connection
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.MutableList
+
+import com.spotify.docker.client.messages.ContainerConfig
+import com.spotify.docker.client._
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.test.SharedSQLContext
+import org.scalatest.BeforeAndAfterAll
+
+abstract class DatabaseOnDocker {
+  /**
+   * The docker image to be pulled
+   */
+  def imageName: String
+
+  /**
+   * A Seq of environment variables in the form of VAR=value
+   */
+  def env: Seq[String]
+
+  /**
+   * jdbcUrl should be a lazy val or a function since `ip` it relies on is only available after
+   * the docker container starts
+   */
+  def jdbcUrl: String
+
+  private val docker: DockerClient = DockerClientFactory.get()
+  private var containerId: String = null
+
+  lazy val ip = docker.inspectContainer(containerId).networkSettings.ipAddress
+
+  def start(): Unit = {
+    while (true) {
+      try {
+        val config = ContainerConfig.builder()
+          .image(imageName).env(env.asJava)
+          .build()
+        containerId = docker.createContainer(config).id
+        docker.startContainer(containerId)
+        return
+      } catch {
+        case e: ImageNotFoundException => retry(5)(docker.pull(imageName))
```

Comment:
Also... why not _first_ try to pull the image and only then try to start the container?